### ISSUE TITLE
SSD reset_class: layer initialization can not be done in GPU context

### ISSUE THREAD

**Issue by lgov** [dmlc/gluon-cv#572]
Using Gluon-CV from the master branch at the current HEAD revision.

When I load a (pre-trained) SSD model, I have the option to load it in a GPU context directly:

```
num_gpus = 2
ctx = [mx.gpu(i) for i in range(num_gpus)]
ssd_resnet50_net = gcv.model_zoo.get_model('ssd_512_resnet50_v1_voc', pretrained=True, ctx=ctx)
```

However, I can not pass the same ctx to the reset_class function:
`
ssd_resnet50_net.reset_class(classes)
`

Reset_class will use the default context (i.e. the CPU). The problem here is that this function creates and initializes new Conv2D layers (the class predictors). 

As a result, later during data loading (on GPU(0) context) I get this error:
`RuntimeError: Parameter 'ssd0_ssd0_convpredictor0_conv0_weight' was not initialized on context gpu(0). It was only initialized on [cpu(0)].`

**Comment by lgov** [comment]
The workaround used in the train_ssd.py example (*) is to load the model in CPU context first, and only at training time reset to the GPU context:

```
149 def train(net, train_data, val_data, eval_metric, ctx, args):
150     """Training pipeline"""
151     net.collect_params().reset_ctx(ctx)
```

If that's considered as the only correct way of loading an SSD model then that's fine for me, but I would drop the 'ctx' parameter from the SSD class to avoid any confusion.

Alternatively, the reset_class method could use the same context as used by the model initialization.

(*) https://gluon-cv.mxnet.io/_downloads/fa6dc3d92420ca2b251af87c0d7fdc90/train_ssd.py

**Comment by zhreshold** [comment]
Either way should be good, but the ctx config is missing for `reset_class`

If you find it's better to have ctx consistently available, feel free to submit a PR for the `reset_class`.

**Comment by zhreshold** [comment]
Agreed, but ctx is not bounded to network itself, but its weights.
it's possible to check the weights in other layers.

**Comment by zhreshold** [comment]
closed by #584
